### PR TITLE
change(web): initializes OSK with keyboard if available during init 🪠

### PR DESCRIPTION
**Pull request by jahorton**
By delaying the end of KeymanWeb's `init` method in order to check if any keyboards are pre-registered, we can initialize the OSK correctly "the first time".  This matters by far the most on touch devices, where we have been building a stand-in default touch-layout the first time first due to not waiting long enough for pre-registered keyboards to fully load.   (On desktop devices, we just build a very simple, empty stand-in.)  Preparing the stand-in is surprisingly expensive/slow on some older touch devices.

By preventing unnecessary construction of a stand-in touch layout (and thus, related DOM initialization + layout reflow) on touch devices when fully leveraged, we can eliminate a notable source of Web engine startup lag that has long been affecting our mobile apps.

I have verified that this allows us to fully bypass the empty touch-layout construction step within the Android Engine, notably cutting the host-page startup time.  It currently appears that the iOS app is unable to leverage this as-is, though.

With a little rework of Web's test pages, it can be demonstrated to work for standalone KeymanWeb as well.  Their current structures generally do not register keyboards until engine-initialization is complete, which is a necessary prerequisite.

## User Testing

TEST_WINDOWS_DESKTOP: Test general use of KeymanWeb on a Windows desktop or laptop. Report any unexpected behavior.
- Refresh the page and start using KeymanWeb least 5 times.  (Test page: "Test unminified Web")

TEST_ANDROID_TABLET: Test general use of KeymanWeb on an Android tablet. Report any unexpected behavior.
- Force-quit and restart the app at least 5 times.
- Swap in and out of the system keyboard at least 5 times.

TEST_IOS_PHONE: Test general use of KeymanWeb on an iPhone. Report any unexpected behavior.
- Force-quit and restart the app at least 5 times.
- Swap in and out of the system keyboard at least 5 times.

### CODE REVIEW

**Comment by jahorton** [comment]
Just noticed an issue with when the OSK is in desktop-keyboard mode; needs a bit more refinement first.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_WINDOWS_DESKTOP (PASSED):** Tested using "Test unminified Web" page on a Windows desktop and here is my observation: 1. Refresh the test page. 2. Tested text selection and CJK keyboard functions on a Windows desktop. 3. Repeated Steps 1 and 2 more than 5 times and verified that everything works fine.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_ANDROID_TABLET (FAILED):** Tested with the attached PR build (Keyman 17.0.303-beta-test-11174)  on an Android 12 Tablet / API 30 emulator and here is my observation:  1. Force-quit and restart the app for 5 times. 2. Swapped in and out of the system keyboard at least 5 times and noticed that the Keyman Keyboard banner appears instead of the predictive text after keyboard swapping. Seems to be an issue.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_IOS_PHONE (FAILED):** Tested with the attached PR build (Keyman 17.0.303-beta-test-11174) on an iPhone 13 mobile device and here is my observation: 1. Noticed that a blank keyboard appeared on the screen after opening the keyman In-App. Seems to be an issue.

**Comment by jahorton** [comment]
@keymanapp-test-bot retest TEST_IOS_PHONE 

@bharanidharanj 

I've fixed the iOS issue thanks to an informative Sentry error report; the keyboard no longer shows up blank on app load.

Also, in regard to the "Android tablet" test, did the wrong-banner behavior occur consistently, or did it only happen occasionally?  I'm having trouble getting a repro.

**Comment by bharanidharanj** [comment]
> @keymanapp-test-bot retest TEST_IOS_PHONE
> 
> @bharanidharanj
> 
> I've fixed the iOS issue thanks to an informative Sentry error report; the keyboard no longer shows up blank on app load.
> 
> Also, in regard to the "Android tablet" test, did the wrong-banner behavior occur consistently, or did it only happen occasionally? I'm having trouble getting a repro.

@jahorton Actually what I did is:
1. Set Keyman as the system keyboard. 
2. Opened Chrome browser. 
3. Swapped to Keyman keyboard. 
4. Typed some text on the Chrome Search bar. 
5. Hold the backspace key to delete all the letters. 
6. Sometimes, I was getting Keyman Crash after holding the backsapce key. 
7. Swapped to Gboard using the Globe key. 
8. Again, swapped to Keyman keyboard. 
9. Repeated step 4 to 8 leads to show the keyman banner on the default EuroLatin (SIL) keyboard.

**Comment by jahorton** [comment]
> 6\. Sometimes, I was getting Keyman Crash after holding the backsapce key.

Digging into Sentry and _also_ getting a repro for myself, I can confirm that _this part_ is due to #11160.  It should thus be fixed by #11188.

I am currently unable to reproduce what's reported for step 9, where the image banner displays instead of the predictive-text banner.

**Comment by bharanidharanj** [comment]
> > 6. Sometimes, I was getting Keyman Crash after holding the backsapce key.
> 
> Digging into Sentry and _also_ getting a repro for myself, I can confirm that _this part_ is due to #11160. It should thus be fixed by #11188.
> 
> I am currently unable to reproduce what's reported for step 9, where the image banner displays instead of the predictive-text banner. Maybe I need to pick the same exact device + API target...

@jahorton I've attached the video file for reference, but reproducing this issue is quite challenging.

https://github.com/keymanapp/keyman/assets/19683143/73342439-87f3-4657-a776-163c3a82c68a

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_IOS_PHONE (PASSED):** Retested with the attached PR build (Keyman 17.0.303-beta-test-11174) on an iPhone 13 Mobile device / iOS 17.4 and here is my observation: 1. Force-quit and restart the app for 5 times. 2. Switched between the system keyboard and another keyboard at least five times and verified that no errors occurred during the process.

**Comment by jahorton** [comment]
> @jahorton I've attached the video file for reference, but reproducing this issue is quite challenging.

That information in itself is useful - we likely have some sort of race condition affecting the banner, then.

There's also a fair shot it _is_ due to work within this PR, given how it affects when the keyboard is loaded and when the check for a matching banner occurs.  I thought I had those bases covered, but this strongly suggests that I should check more thoroughly.

I thought I saw a similar effect in a different PR recently... but I can't find evidence of that now that I look.  (If it _did_ happen elsewhere as well, that would argue against it being from this PR.)

**Comment by jahorton** [comment]
@keymanapp-test-bot retest TEST_IOS_PHONE TEST_ANDROID_TABLET

I've added code that should ensure the banner displays consistently now; the OSK now proactively checks what mode the banner should be in when it's initialized.  (Previously, it always waited to be told.)

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_ANDROID_TABLET (PASSED):** Retested with the attached PR build (Keyman 17.0.303-beta-test-11174) on an Android 12 / API 30 emulator and here is my observation: 1. Opened the Keyman In-app. 2. Force-quit and restarted the app more than five times, verifying that no issues occurred during this process. 3. Set Keyman as the System keyboard. 4. Opened Chrome browser. 5. Switched to Keyman Keyboard. 6. Swapped to Gboard using the globe key. Repeat step 5 and 6 for more than five times and verified that no problems occurred during the process.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_IOS_PHONE (PASSED):** Retested with the attached PR build (Keyman 17.0.303-beta-test-11174) on an iPhone 13 Mobile device (iOS 17.4) and here is my observation: 1. Opened the Keyman In-app. 2. Force-quit and restarted the app more than five times, verifying that no issues occurred during this process. 3. Set Keyman as the System keyboard. 4. Opened Safari browser. 5. Switched to Keyman Keyboard. 6. Swapped to default keyboard using the globe key. Repeat step 5 and 6 for more than five times and verified that no problems occurred during the process.

**Review by ermshiperete — APPROVED** [review]

File: android/KMEA/app/src/main/assets/android-host.js [hunk 1]
```diff
@@ -36,7 +36,9 @@ function init() {
   keyman.beepKeyboard = beepKeyboard;
 
   // Readies the keyboard stub for instant loading during the init process.
-  KeymanWeb.registerStub(JSON.parse(jsInterface.initialKeyboard()));
+  try {
+    KeymanWeb.registerStub(JSON.parse(jsInterface.initialKeyboard()));
+  } catch {};
```
[the remark this answers is not in the document]
Comment (jahorton) [reply android/KMEA/app/src/main/assets/android-host.js:41]: Note:  this is location that, if an error triggers, is _guaranteed_ to cause a blank keyboard.  It's before the `keyman.init` call.

Here's a Sentry event this is meant to silence: https://keyman.sentry.io/share/issue/d9d9bd46c7404254891b606b87570bec/

I'm... not sure how that error looks on the JS side.  I can attempt to reproduce it and see if there's a simple-enough way to filter it.

[the remark this answers is not in the document]
Comment (mcdurdin) [reply android/KMEA/app/src/main/assets/android-host.js]: Perhaps the best solution for now is log all exceptions to sentry, and then we can filter out known 'good' exceptions as we encounter them. So we should never have an empty `catch {}` -- it should at the very least have a `tellSentryAboutMyPain()` even if we then want to continue on as 'normal'.

[the remark this answers is not in the document]
Comment (jahorton) [reply android/KMEA/app/src/main/assets/android-host.js]: I went ahead and dropped the try-catch entirely.

[the remark this answers is not in the document]
Comment (jahorton) [reply android/KMEA/app/src/main/assets/android-host.js]: ... and restored, now with a `console.error(<error>)` in it.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.311-beta](https://keyman.com/downloads/releases/beta/17.0.311)

**Pull request merged**